lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,885 @@
+    // Register the source operation with the current line id
+    size_t opLID = getOpLID(srcop);
+
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Build and return the state instruction
+    os << opLID << " "
+       << "state"
+       << " " << sid << " " << name << "\n";
+  }
+
+  // Generates a next instruction, given a width, a state LID, and a next value
+  // LID
+  void genNext(Value next, Operation *reg, int64_t width) {
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Retrieve the LIDs associated to reg and next
+    size_t regLID = getOpLID(reg);
+    size_t nextLID = getOpLID(next);
+
+    // Build and return the next instruction
+    // Also update the lid as this instruction is not associated to an mlir op
+    os << lid++ << " "
+       << "next"
+       << " " << sid << " " << regLID << " " << nextLID << "\n";
+  }
+
+  // Verifies that the sort required for the given operation's btor2 emission
+  // has been generated
+  int64_t requireSort(mlir::Type type) {
+    // Start by figuring out what sort needs to be generated
+    int64_t width = hw::getBitWidth(type);
+
+    // Sanity check: getBitWidth can technically return -1 it is a type with no
+    // width (like a clock). This shouldn't be allowed as width is required to
+    // generate a sort
+    assert(width != noWidth);
+
+    // Generate the sort regardles of resulting width (nothing will be added if
+    // the sort already exists)
+    genSort("bitvec", width);
+    return width;
+  }
+
+  // Generates the transitions required to finalize the register to state
+  // transition system conversion
+  void finilizeRegVisit(Operation *op) {
+    // Check the register type (done to support non-firrtl registers as well)
+    auto reg = dyn_cast<seq::FirRegOp>(op);
+    if (!reg)
+      return;
+
+    // Generate the reset condition (for sync & async resets)
+    // We assume for now that the reset value is always 0
+    size_t width = hw::getBitWidth(reg.getType());
+    genSort("bitvec", width);
+
+    // Extract the `next` operation for each register (used to define the
+    // transition). We need to check if next is a port to avoid nullptrs
+    Value next = reg.getNext();
+
+    // Next should already be associated to an LID at this point
+    // As we are going to override it, we need to keep track of the original
+    // instruction
+    size_t nextLID = noLID;
+
+    // Check for special case where next is actually a port
+    // To do so, we start by checking if our operation is a block argument
+    if (BlockArgument barg = dyn_cast<BlockArgument>(next)) {
+      // Extract the block argument index and use that to get the line number
+      size_t argIdx = barg.getArgNumber();
+
+      // Check that the extracted argument is in range before using it
+      if (inputLIDs.contains(argIdx)) {
+        nextLID = inputLIDs[argIdx];
+      }
```

Comment:
you're right, while the case should never happen, I did still leave it there out of force of habit.

Reply:
I think that originally I did this because the clock counted as an argument and I didn't want it to be considered (as the clock is completely ignored in the output btor2.